### PR TITLE
Tdl 14376 pagination failure

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
- Updated one condition to break out of pagination loop.
  - If a whole blank page found, then stop looping.
- Fixed pagination test case.

Note : Unittest case was not possible for this PR. We have replaced just one condition and  method which having this condition is too large and using lot of other methods. There is sync method with sync.py module. That's why to mock the other methods was not possible due to same name of method and module. So, skipped it.
# Manual QA steps
 - Verify that pagination does not break if last 2 records of any page is empty.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_google_sheets_pagination.py [hunk 1]
```diff
@@ -61,6 +61,34 @@ def test_run(self):
...
+                    primary_keys_page_1 = set(primary_keys_list_1)
+                    primary_keys_page_2 = set(primary_keys_list_2)
+
+                    # Verify by primary keys that data is unique for page
+                    self.assertTrue(
+                        primary_keys_page_1.isdisjoint(primary_keys_page_2))
+
+                if stream == "sadsheet-pagination":
+                    # verify the data for the "sadsheet-pagination" stream is free of any duplicates or breaks by checking
+                    # our fake pk value ('id')
```
Comment (savan-chovatiya) [inline tests/test_google_sheets_pagination.py:79]: Add a detailed comment about the data present in this sheet.

Comment (prijendev) [reply tests/test_google_sheets_pagination.py]: Added

**Review by savan-chovatiya — APPROVED** [review]

File: tap_google_sheets/sync.py [hunk 2]
```diff
@@ -514,7 +514,7 @@ def sync(client, config, catalog, state):
                             from_row=from_row,
                             columns=columns,
                             sheet_data_rows=sheet_data_rows)
-                        if row_num < to_row:
+                        if not sheet_data_rows: # If a whole blank page found, then stop looping.
```
Comment (karanpanchal-crest) [inline tap_google_sheets/sync.py:517]: @prijendev Can you please explain what the earlier behavior `row_num < to_row:` meant?

Comment (prijendev) [reply tap_google_sheets/sync.py:517]: Here, `to_row` is initialized with a minimum of 200(max page size) or `max_row`. Then, it continues to add 200 until `max_row`.  Initially, `from_row` is assigned by 2, and from the next page, it is assigned by `to_row`+1. `row_num` is the addition of `from_row` and total records get in response.  The above condition checks that if row_num is less than to_row or not based on which it decides set is_last_row true. But API does not return the last empty rows in response. 

For example, rows 199 and 200 are empty, and a total 400 rows are there in the sheet. So, in 1st iteration

`to_row` = 200
`from_row` = 2
`row_num` = 2 + 197 = 199(1st row contain header value)
So, the above condition becomes true and breaks the loop.

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KrisPersonal — CHANGES_REQUESTED** [review]
1) Please add comments in the code as to why if row_num < to_row is replaced with sheet_data_rows.
2) If for example, if a sheet contains only 200 rows and let us say 99 and 100 are empty, then it will continue to process the remaining rows because of this condition being removed and by adding sheet_data_rows empty condition it will check whether the whole sheet is empty. Is that correct?

**Comment by prijendev** [comment]
> 1. Please add comments in the code as to why if row_num < to_row is replaced with sheet_data_rows.
> 2. If for example, if a sheet contains only 200 rows and let us say 99 and 100 are empty, then it will continue to process the remaining rows because of this condition being removed and by adding sheet_data_rows empty condition it will check whether the whole sheet is empty. Is that correct?



> 1. Please add comments in the code as to why if row_num < to_row is replaced with sheet_data_rows.
> 2. If for example, if a sheet contains only 200 rows and let us say 99 and 100 are empty, then it will continue to process the remaining rows because of this condition being removed and by adding sheet_data_rows empty condition it will check whether the whole sheet is empty. Is that correct?

> 1. Added detail comments in the code. 
> 2. sheet_data_rows will check for an empty page, not a whole empty sheet.

File: tests/test_google_sheets_pagination.py [hunk 3]
```diff
@@ -43,9 +44,8 @@ def test_run(self):
         record_count_by_stream = self.run_and_verify_sync(conn_id)
         synced_records = runner.get_records_from_target_output()
 
-        for stream in testable_streams.difference({
-                'sadsheet-pagination',  # BUG TDL-14376
-        }):
+        # Added back `sadsheet-pagination` to testable_streams as # BUG TDL-14376 resolved.
```
Comment (kspeer825) [inline tests/test_google_sheets_pagination.py:47]: This comment isn't necessary.

File: tests/test_google_sheets_pagination.py [hunk 4]
```diff
@@ -61,6 +61,43 @@ def test_run(self):
                 # verify that we can paginate with all fields selected
                 self.assertGreater(record_count_by_stream.get(stream, 0), self.API_LIMIT)
 
+                record_count_sync = record_count_by_stream.get(stream, 0)
```
Comment (kspeer825) [inline tests/test_google_sheets_pagination.py:63]: There should not have needed to be any test changes here besides adding the failing sheet back to the test. I don't think these test additions are adding any test coverage. The sheets have been setup in a way that relies on a specific column with incrementing values used to compare against the sdc column. See fake_pk_list above.

File: tests/test_google_sheets_pagination.py (hunk 3 above)
Comment (prijendev) [reply tests/test_google_sheets_pagination.py]: Removed it.

**Review by kspeer825 — CHANGES_REQUESTED** [review]
We are striving to writeup defects found by QA in a way that makes reproducing issues as simple as uncommenting/commenting lines marked with BUG. Unless there is an explicit TODO left in the test, or a bullet in the DoD of the card there shouldn't be any missing test cases. I believe that is the case here and the test additions can be removed. If you find this is not the case please raise this to us so we can adjust our bug reporting process.

File: tests/test_google_sheets_pagination.py (hunk 4 above)
Comment (prijendev) [reply tests/test_google_sheets_pagination.py]: Removed extra assertion. The current test is written in such a way that it was respecting only `Pagination` stream. 
https://github.com/singer-io/tap-google-sheets/blob/25136fce609ca297256d63c2da2112fe2512ebae/tests/test_google_sheets_pagination.py#L80-L87
If we add back `sadsheet-pagination` then we have to write assertion according to this sheet also.

Comment (kspeer825) [reply tests/test_google_sheets_pagination.py]: That comment is misleading, the test is setup to apply to both sheets
https://github.com/singer-io/tap-google-sheets/blob/25136fce609ca297256d63c2da2112fe2512ebae/tests/test_google_sheets_pagination.py#L38
Additionally there is a large comment at the top of this test that should be removed https://github.com/singer-io/tap-google-sheets/blob/25136fce609ca297256d63c2da2112fe2512ebae/tests/test_google_sheets_pagination.py#L9-L13

Comment (prijendev) [reply tests/test_google_sheets_pagination.py]: In the `sadsheet-pagination` sheet, there is a total of 238 rows and in that sheet row, no 199 and 200 are empty rows whereas in the `Pagination` sheet there are total of 239 rows with no empty row.
So, as two rows are empty in `sadsheet-pagination`, we need to write the separate assertion in which we are excluding rows no 199 and 200.

**Review by KrisPersonal — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**